core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1356,6 +1373,78 @@ class KafkaApis(val requestChannel: RequestChannel,
       ))
   }
 
+  def handleDescribeTopicPartitionsRequest(request: RequestChannel.Request): Unit = {
+    metadataCache match {
+      case _: ZkMetadataCache =>
+        throw new InvalidRequestException("ZK cluster does not handle DescribeTopicPartitions request")
+      case _ =>
+    }
+    val kRaftMetadataCache = metadataCache.asInstanceOf[KRaftMetadataCache]
+
+    val describeTopicPartitionsRequest = request.body[DescribeTopicPartitionsRequest].data()
+    val topicSet = mutable.SortedSet[String]()
+    describeTopicPartitionsRequest.topics().forEach(topic => topicSet.add(topic.name()))
+    val topics = ListBuffer[String]().addAll(topicSet)
+
+    val cursor = describeTopicPartitionsRequest.cursor()
+    val fetchAllTopics = topics.isEmpty
+    if (fetchAllTopics) {
+      kRaftMetadataCache.getAllTopics().foreach(topic => topics.append(topic))
```

Comment:
If we copy and sort all the topic names anyway, do we need to change the underlying data structure to NavigableMap?  We could just use this list to traverse topic info and it will be in order.

Reply:
In the fetch all path, no additional sort is required. I did not see a good way to convert Java list to a scala mutable list, so I did the copy.
Use a mutable list for 2 reasons
1. It is easier to filter out the topics alphabetically ahead of the cursor topic
2. In the fetch all case, I think we should still include the cursor topic in the response if it does not exist. Mutable list make it easier.

Reply:
But if you ask whether it is worth the effort to create the full set of underline structures to get an ordered list where we can just sort the list, I am not sure.